software/ivar/consensus/main.nf

```diff
@@ -0,0 +1,46 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName } from './functions'
+
+params.options = [:]
+def options    = initOptions(params.options)
+
+process IVAR_CONSENSUS {
+    tag "$meta.id"
+    label 'process_medium'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), publish_id:meta.id) }
+
+    conda (params.enable_conda ? "bioconda::ivar=1.3.1" : null)
+
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/ivar:1.3.1--h089eab3_0"
+    } else {
+        container "quay.io/biocontainers/ivar:1.3.1--h089eab3_0"
+    }
+
+    input:
+    tuple val(meta), path(bam)
+
+    output:
+    tuple val(meta), path("*.fa")      , emit: cns_seq
+    tuple val(meta), path("*.qual.txt"), emit: cns_qual
+    tuple val(meta), path("*_mpileup.txt"), emit: mpileup
+    path "*.version.txt"               , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    """
+    samtools mpileup \\
+        -aa -A -d 0 -Q 0 \\
```

Comment:
```suggestion
```

Reply:
These options should come in via `options.args2` for flexibility.

Reply:
That is fair. But, they do seem to the recommended options from the ivar manual to run it, which is why I put them here. As these would be considered best practice, and suggested by the developer, should they remain? I am unclear what effect not having these exact parameters would have on the outcome of ivar consensus. I will test it out.

Reply:
These options have to come in via `options.args` because the emphasis for these modules is for anyone to be able to use them with pretty much their own parameter set. It would also allow testing of different parameters directly via the pipeline as you suggested you would do. Ironically, those parameters are different in different examples in the [iVar manual](https://andersen-lab.github.io/ivar/html/manualpage.html) 😆

```
Usage: samtools mpileup -aa -A -d 0 -B -Q 0 --reference [<reference-fasta] <input.bam> | ivar variants -p <prefix> [-q <min-quality>] [-t <min-frequency-threshold>] [-m <minimum depth>] [-r <reference-fasta>] [-g GFF file]
Note : samtools mpileup output must be piped into ivar variants
```

```
Example Usage:

samtools mpileup -aa -A -d 600000 -B -Q 0 test.trimmed.bam | ivar variants -p test -q 20 -t 0.03 -r test_reference.fa -g test.gff
```

This reminds me. We need to add `path fasta` as an `input:` to the module too and add it to the `samtools mpileup` command via `--fasta-ref $fasta`.

Reply:
Cool. Thank you @drpatelh. I was thinking of adding the whole string as a default option to args.opts2, but I think you suggestion is far more practical.